### PR TITLE
gh-114123: migrate docstring from _csv to csv

### PR DESCRIPTION
**Pull request by smontanaro**
* Issue: gh-114123

### CODE REVIEW

**Comment by smontanaro** [comment]
I don't think this requires a News entry. In theory, the `_csv` module is an implementation detail. Nothing changed when asking for `csv.__doc__`.

**Comment by smontanaro** [comment]
Not sure who the default documentation reviewer is. I suspect @encukou will know...

File: Lib/csv.py
Comment (AA-Turner) [inline Lib/csv.py]: We need to make this a raw string as otherwise `help(csv)` will show:

```
[...snip...]
DESCRIPTION
[...snip...]
       class excel:
            delimiter = ','
            quotechar = '"'
            escapechar = None
            doublequote = True
            skipinitialspace = False
            lineterminator = '
    '
            quoting = QUOTE_MINIMAL
```

```suggestion
r"""
```

**Review by AA-Turner — APPROVED** [review]
This seems fine but I'm not a ``csv`` expert!

A

File: Lib/csv.py
Comment (AA-Turner) [inline Lib/csv.py]:
```suggestion
    * lineterminator - specifies the character sequence which should
```

File: Lib/csv.py
Comment (smontanaro) [reply Lib/csv.py]: Thanks!

File: Lib/csv.py
Comment (smontanaro) [reply Lib/csv.py]: Thanks!

**Comment by AA-Turner** [comment]
Thoughts on backports? I'm somewhat tempted, as future changes to the CSV docstring will be much harder if this PR isn't backported.

A

**Comment by smontanaro** [comment]
> Thoughts on backports? I'm somewhat tempted, as future changes to the CSV
> docstring will be much harder if this PR isn't backported.
>

I don't really have an opinion on the topic. If it simplifies future edits
(are doc changes common in older versions?) I guess it's fine.

That said, I also have no idea what's involved to backport the change.

>

**Review by serhiy-storchaka — APPROVED** [review]
I think that this change needs a NEWS entry, because it has a user visible effect in public module (for example removing `"__doc__"` from `__all__`).

And while we are here, why not remove also `"__version__"` from `__all__`?

**Comment by smontanaro** [comment]
@serhiy-storchaka Thanks, I'll take care of both later today.

**Comment by erlend-aasland** [comment]
@serhiy-storchaka, some code may rely on `__version__`. I think this PR should focus on the docstring only. See previous more or less relevant discussions:

- https://mail.python.org/archives/list/python-dev@python.org/thread/KBU4EU2JULXSMUZULD5HJJWCGOMN52MK/#W2VCBOOS7NSUMEDBO7UOB22Z3CBVZIV5
- https://github.com/python/cpython/issues/76007
- https://github.com/python/cpython/issues/93370

**Comment by smontanaro** [comment]
I didn't see @erlend-aasland's note about creeping `__version__`ism. That said, I'm happy to remove it from the `csv` module altogether if that's what folks want. Maybe someone relies on it, but I'd be kinda surprised. In any case, it doesn't belong in `__all__` (IMO), and there's no need to define it in C code just to import it into the front-end Python module.

Removing `__version__` shouldn't be backported though. If removing it is the path forward, I should probably do it in a second PR which isn't going to be backported.

File: Misc/NEWS.d/next/Documentation/2024-01-17-11-40-03.gh-issue-114123.LuueXf.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Documentation/2024-01-17-11-40-03.gh-issue-114123.LuueXf.rst]: `Modules/_csv.c` and `Lib/csv.py` are not meaningful names for common Python user. It is better to write about `csv` and `_csv` modules.

Also not that it is reStructuredText, not Markdown. Use ``` :mod:`csv` ``` if you want to create a reference and ``` :mod:`!csv` ``` or just ``` ``csv`` ``` if do not.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM, besides niypicks about a NEWS file.

File: Misc/NEWS.d/next/Documentation/2024-01-17-11-40-03.gh-issue-114123.LuueXf.rst
Comment (AA-Turner) [inline Misc/NEWS.d/next/Documentation/2024-01-17-11-40-03.gh-issue-114123.LuueXf.rst]: Building on Serhiy's notes:
```suggestion
Move the :mod:`csv` module docstring to the :mod:`!csv` module
instead of reexporting it from the internal :mod:`!_csv` module,
and remove ``__doc__`` from ``csv.__all__``.

Move :attr:`!csv.__version__` to the :mod:`!csv` module
instead of reexporting it from the internal :mod:`!_csv` module,
and remove ``__version__`` from ``csv.__all__``.
```

Comment (smontanaro) [reply Misc/NEWS.d/next/Documentation/2024-01-17-11-40-03.gh-issue-114123.LuueXf.rst]: Done.

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_csv.py
Comment (AA-Turner) [inline Lib/test/test_csv.py]: I think `test__all__` should be kept?


```suggestion
    def test__all__(self):
        support.check__all__(self, csv, {'csv', '_csv'})
```

Comment (smontanaro) [reply Lib/test/test_csv.py]: @AA-Turner Isn't `test__all__` a standard thing? Why is this override needed if we no longer have an `extra`?

**Comment by smontanaro** [comment]
Before anybody merges (not my job), let's settle on the fate of `csv.__version__`.

File: Lib/test/test_csv.py
Comment (AA-Turner) [reply Lib/test/test_csv.py]: ``test.support.check__all__()`` is, but ``test__all__`` isn't defined in ``unittest.TestCase``, so we do need to define the test to keep it.

**Comment by AA-Turner** [comment]
No strong feelings on removing `csv.__version__`, but I think it should be done in a new PR with a proper deprecation, which wouldn't block this PR's goal of moving ``__version__`` to ``csv.py`` (and indeed, having ``__version__`` defined in Python would make the deprecation mildly easier).

```python
_deprecated___version__ = "1.0"

def __getattr__(name):
    if name == "__version__":
        from warnings import _deprecated

        _deprecated(name, remove=(3, 15))
        return globals()[f"_deprecated_{name}"]
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

A

File: Lib/test/test_csv.py
Comment (serhiy-storchaka) [reply Lib/test/test_csv.py]: No, it is not overridden, it is added. It compares the `__all__` content with the calculated expected value (all public function, classes, etc defined in the specified modules). `extra` is an extra.

**Comment by erlend-aasland** [comment]
The fate of `__version__` lies in another issue and PR. We can use #76007 for the issue, as *csv* is already mentioned there.

**Comment by smontanaro** [comment]
> No objection to removing `csv.__version__`, but I think it should be done in a new PR with a proper deprecation, which wouldn't block this PR's goal of moving `__version__` to `csv.py` (and indeed, having `__version__` defined in Python would make the deprecation mildly easier).
> 
> ```python
> _deprecated___version__ = "1.0"
> 
> def __getattr__(name):
>     if name == "__version__":
>         from warnings import _deprecated
> 
>         _deprecated(name, remove=(3, 15))
>         return globals()[f"_deprecated_{name}"]
>     raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
> ```
> 
> A

Is this the standard idiom for deprecating names?

**Comment by erlend-aasland** [comment]
> Is this the standard idiom for deprecating names?

That pattern (and variations over it) is used several places in `Lib/`, so I guess you can call it a standard idiom.

File: Lib/csv.py
Comment (merwok) [inline Lib/csv.py]:
```suggestion
CSV parsing and writing.
```

File: Lib/csv.py
Comment (merwok) [inline Lib/csv.py:6]:
```suggestion
of comma-separated values files, and implements the interface
```

**Comment by smontanaro** [comment]
> Is this the standard idiom for deprecating names?
>
> That pattern (and variations over it) is used several places in Lib/, so
> I guess you can call it a standard idiom.
>

Any possibility of standardizing it enough to stay out somewhere in the std
lib? (Just thinking out loud. After a few times, it seems like it would be
useful to make it into a battery.)

>

**Comment by merwok** [comment]
I am not sure that there’s enough demand for it, nor a natural place.
It’s too specific to the Python stdlib to be in `warnings`, and I don’t think there’s any module concerned with meta-stdlib utilities.

**Comment by smontanaro** [comment]
I don't think this change should be backported. How did the attempt happen?

**Comment by AA-Turner** [comment]
I added the backport labels after my note [above](https://github.com/python/cpython/pull/114124#issuecomment-1894592735):

> Thoughts on backports? I'm somewhat tempted, as future changes to the CSV docstring will be much harder if this PR isn't backported

Though given the backport would require manual work regardless, I agree on leaving it for now & not backporting.

A

**Comment by smontanaro** [comment]
Hmmm... Okay, go for it. Let the chips fall where they may.

**Comment by merwok** [comment]
Even if this is all about internal location of attributes, it feels safer to not backport.
Not worth any risk, and possible docstring backports will have to be done by hand.

**Pull request merged**
